zookeeper/source/zk_nn_client.cc

```diff
@@ -0,0 +1,952 @@
+		memcpy((&blockId), &uuid, sizeof(u_int64_t) / sizeof(u_char));
+		return true;
+	}
+
+	// TODO: To simplify signature, could just get rid of the newBlock param
+	// and always check for preexisting replicas
+	bool ZkNnClient::find_datanode_for_block(std::vector<std::string>& datanodes, const u_int64_t blockId, uint32_t replication_factor, bool newBlock) {
+		// TODO: Actually perform this action
+		// TODO: Perhaps we should keep a cached list of nodes
+
+		std::vector<std::string> live_data_nodes = std::vector <std::string>();
+		int error_code;
+
+		// Get all of the live datanodes
+		if (zk->get_children(HEALTH, live_data_nodes, error_code)) {
+
+			// LOG(INFO) << CLASS_NAME << "Found live DNs: " << live_data_nodes;
+			auto excluded_datanodes = std::vector <std::string>();
+			if (!newBlock) {
+				// Get the list of datanodes which already have a replica
+				if (zk->get_children(BLOCK_LOCATIONS + std::to_string(blockId), excluded_datanodes, error_code)) {
+					// Remove the excluded datanodes from the live list
+					for (auto excluded : excluded_datanodes) {
+						std::vector<std::string>::iterator it = std::find(
+								live_data_nodes.begin(),
+								live_data_nodes.end(),
+								excluded);
+
+						if (it != live_data_nodes.end()) {
+							// The excluded dn was found, so remove it
+							live_data_nodes.erase(it);
+						}
+					}
+				} else {
+					LOG(ERROR) << CLASS_NAME << "Error getting children of: " << BLOCK_LOCATIONS + std::to_string(blockId);
+					return false;
+				}
+			}
+			// LOG(INFO) << CLASS_NAME << "Live DNs after excluding those with existing replicas: " << live_data_nodes;
+
+			/* for each child, check if the ephemeral node exists */
+			for(auto datanode : live_data_nodes) {
+				bool isAlive;
+				if (!zk->exists(HEALTH_BACKSLASH + datanode + HEALTH, isAlive, error_code)) {
+					LOG(ERROR) << CLASS_NAME << "Failed to check if datanode: " + datanode << " is alive: " << error_code;
+				}
+				if (isAlive) {
+					datanodes.push_back(datanode);
+				}
+				if (datanodes.size() == replication_factor) {
+					LOG(INFO) << CLASS_NAME << "Found " << replication_factor << " datanodes";
+					break;
+				}
+			}
+		} else {
+			LOG(ERROR) << CLASS_NAME << "Failed to get list of datanodes at " + HEALTH + " " << error_code;
+			return false;
+		}
+
+		// TODO: Read strategy from config, but as of now select the first few blocks that are valid
+
+		if (datanodes.size() > replication_factor) {
+			LOG(ERROR) << CLASS_NAME << "Failed to find at least " << replication_factor << " datanodes at " + HEALTH;
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Updates /fileSystem in ZK to reflect a file rename
+	 * @param src The path to the source file (not znode) within the filesystem
+	 * @param dst The path to the renamed destination file (not znode) within the filesystem
+	 * @return Boolean indicating success or failure of the rename
+	 */
+	bool ZkNnClient::rename_file(std::string src, std::string dst) {
```

Comment:
Do we have to perform any locking on this file to guarantee no racy conditions while copying?